### PR TITLE
[SMTChecker] Make smt::Sort::operator== virtual

### PR DESCRIPTION
**Pull request by leonardoalt**

### CODE REVIEW

File: libsolidity/formal/SolverInterface.h [hunk 1]
```diff
@@ -65,16 +65,17 @@ struct FunctionSort: public Sort
 {
 	FunctionSort(std::vector<SortPointer> _domain, SortPointer _codomain):
 		Sort(Kind::Function), domain(std::move(_domain)), codomain(std::move(_codomain)) {}
-	bool operator==(FunctionSort const& _other) const
+	bool operator==(Sort const& _other) const override
 	{
+		auto _otherFunction = dynamic_cast<FunctionSort const&>(_other);
```
Comment (axic) [inline libsolidity/formal/SolverInterface.h]: We usually put `solAssert` before such casts to have the usual detailed exception.

File: libsolidity/formal/SolverInterface.h [hunk 2]
```diff
@@ -87,9 +88,10 @@ struct ArraySort: public Sort
 	/// _range is the sort of the values
 	ArraySort(SortPointer _domain, SortPointer _range):
 		Sort(Kind::Array), domain(std::move(_domain)), range(std::move(_range)) {}
-	bool operator==(ArraySort const& _other) const
+	bool operator==(Sort const& _other) const override
 	{
-		return Sort::operator==(_other) && *domain == *_other.domain && *range == *_other.range;
+		auto _otherArray = dynamic_cast<ArraySort const&>(_other);
```
Comment (axic) [inline libsolidity/formal/SolverInterface.h]: Same here.

File: libsolidity/formal/SolverInterface.h (hunk 1 above)
Comment (leonardoalt) [reply libsolidity/formal/SolverInterface.h]: Even with references? This way the cast would throw an exception anyway

**Comment by leonardoalt** [comment]
@axic Updated

File: libsolidity/formal/SolverInterface.h [hunk 3]
```diff
@@ -65,16 +65,18 @@ struct FunctionSort: public Sort
 {
 	FunctionSort(std::vector<SortPointer> _domain, SortPointer _codomain):
 		Sort(Kind::Function), domain(std::move(_domain)), codomain(std::move(_codomain)) {}
-	bool operator==(FunctionSort const& _other) const
+	bool operator==(Sort const& _other) const override
 	{
+		auto _otherFunction = dynamic_cast<FunctionSort const*>(&_other);
+		solAssert(_otherFunction, "");
```
Comment (chriseth) [inline libsolidity/formal/SolverInterface.h:73]: Shouldn't it just return false if the other is not a function?

Comment (chriseth) [reply libsolidity/formal/SolverInterface.h:73]: Or start the function with `Sort::operator==(_other)`.

Comment (leonardoalt) [reply libsolidity/formal/SolverInterface.h:73]: Yea, that makes sense

**Comment by leonardoalt** [comment]
Updated

File: libsolidity/formal/SolverInterface.h [hunk 4]
```diff
@@ -65,16 +65,20 @@ struct FunctionSort: public Sort
...
 	FunctionSort(std::vector<SortPointer> _domain, SortPointer _codomain):
 		Sort(Kind::Function), domain(std::move(_domain)), codomain(std::move(_codomain)) {}
-	bool operator==(FunctionSort const& _other) const
+	bool operator==(Sort const& _other) const override
 	{
+		if (!Sort::operator==(_other))
+			return false;
+		auto _otherFunction = dynamic_cast<FunctionSort const*>(&_other);
+		solAssert(_otherFunction, "");
 		if (!std::equal(
```
Comment (chriseth) [inline libsolidity/formal/SolverInterface.h:76]: You need to check the size to be equal. I would recommend rather using a boost function that does this properly.

Comment (leonardoalt) [reply libsolidity/formal/SolverInterface.h:76]: I'd rather check it explicitly if that's also fine. Updated.

**Review by chriseth — APPROVED** [review]

**Pull request merged**